nestkernel/conn_builder_conngeninterface.cpp

```diff
@@ -20,158 +20,133 @@
+    while ( cg_->next( source, target, NULL ) )
     {
       // No need to check for locality of the target, as the mask
-      // created by cg_set_masks() only contain local nodes.
-      kernel().connection_manager.connect(
-        ( *source_node_ids )[ source ], ( *target_node_ids )[ target ], dummy_params, synmodel_id );
+      // created by cg_set_masks() only contains local nodes.
+      Node* const target_node = kernel().node_manager.get_node_or_proxy( ( *targets_ )[ target ] );
+      const thread target_thread = target_node->get_thread();
+      single_connect_( ( *sources_ )[ source ], *target_node, target_thread, rng );
     }
   }
   else if ( num_parameters == 2 )
   {
-    if ( not params_map->known( names::weight ) or not params_map->known( names::delay ) )
+    if ( not params_map_->known( names::weight ) or not params_map_->known( names::delay ) )
     {
       throw BadProperty( "The parameter map has to contain the indices of weight and delay." );
     }
 
-    const size_t w_idx = ( *params_map )[ names::weight ];
-    const size_t d_idx = ( *params_map )[ names::delay ];
+    const size_t d_idx = ( *params_map_ )[ names::delay ];
+    const size_t w_idx = ( *params_map_ )[ names::weight ];
 
-    const bool w_idx_is_0_or_1 = ( w_idx == 0 ) or ( w_idx == 1 );
     const bool d_idx_is_0_or_1 = ( d_idx == 0 ) or ( d_idx == 1 );
+    const bool w_idx_is_0_or_1 = ( w_idx == 0 ) or ( w_idx == 1 );
     const bool indices_differ = ( w_idx != d_idx );
-    if ( not( w_idx_is_0_or_1 and d_idx_is_0_or_1 and indices_differ ) )
+    if ( not( d_idx_is_0_or_1 and w_idx_is_0_or_1 and indices_differ ) )
     {
-      throw BadProperty( "w_idx and d_idx have to differ and be either 0 or 1." );
+      throw BadProperty( "d_idx and w_idx have to be either 0 or 1 and cannot be the same." );
     }
 
     std::vector< double > params( 2 );
 
+    const thread num_threads = kernel().vp_manager.get_num_threads();
+
     // connect source to target with weight and delay
-    while ( cg->next( source, target, &params[ 0 ] ) )
+    while ( cg_->next( source, target, &params[ 0 ] ) )
     {
       // No need to check for locality of the target node, as the mask
-      // created by cg_set_masks() only contain local nodes.
-      Node* const target_node = kernel().node_manager.get_node_or_proxy( ( *target_node_ids )[ target ] );
+      // created by cg_set_masks() only contains local nodes.
+      Node* const target_node = kernel().node_manager.get_node_or_proxy( ( *targets_ )[ target ] );
       const thread target_thread = target_node->get_thread();
 
-      kernel().connection_manager.connect( ( *source_node_ids )[ source ],
-        target_node,
-        target_thread,
-        synmodel_id,
-        dummy_params,
-        params[ d_idx ],
-        params[ w_idx ] );
+      delay_ = new ScalarDoubleParameter( params[ d_idx ], num_threads );
+      weight_ = new ScalarDoubleParameter( params[ w_idx ], num_threads );
```

Comment:
@hakonsbm: I was hoping that setting `weight_` and `delay_` here would have the effect that the subsequent call to `single_connect` will just use these values for establishing the connections. Unfortunately, this seems to not be the case and the connections are just created with defaults.

A completely orthogonal idea would of course be to directly call `ConnectionManager::connect()` here and not use any `Parameters` and `single_connect()` at all. For that I'd probably would need to get some values (e.g. the synapse model) from the *syn_spec*.

So, the point is that, the whole mechanism for `Parameters` and how to implement a `ConnBuilder` has apparently become a bit too complex for me and I'm a bit at loss here. Could you maybe give me some advice? I suggest to have a dedicated session on this during the hackathon next week in which we could also discuss a bit about how to document all this.